### PR TITLE
Add Boilerplate Code for Session 2

### PR DESCRIPTION
**Pull request by mrdivinemaniac**
**Description**
This adds some boilerplate code for session 2. All of this code has already been introduced in Session I.
The addition of all this boilerplate code before the session saves some time in the actual session.

This also changes the server api spec.
- `/login` is now `/auth/token`
- `/logout` is now `/auth/revoke`
- Adds `/signup` to the spec

Code that is yet to be implemented:
- Implement `GET /posts`, `GET /posts/:id`, GET `PATCH /posts/:id` and `DELETE /posts/:id`
- JWT creation (`signToken`)
- JWT verification (`verifyToken`, `checkAuth`)
- Code dealing with login tokens in the CLI (`saveToken`, `readToken`, `deleteToken`)

### CODE REVIEW

File: cli/modules/login/index.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      name: {
+        required: true
+      },
+      password: {
+        required: true,
+        hidden: true
+      }
+    }
+  })
+}
```
Comment (mrdivinemaniac) [inline cli/modules/login/index.js:39]: This part should be pretty self explanatory but will be explained in the session.

File: cli/utils/index.js [hunk 2]
```diff
@@ -3,6 +3,20 @@ const path = require('path')
...
+}
+
+function readToken () {
+  
+}
+
+function deleteToken () {
+  
+}
+
```
Comment (mrdivinemaniac) [inline cli/utils/index.js]: We will be implementing these in the session

File: server/middleware/index.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+function checkAuth () {
+  return async function checkAuthMiddleware (req, res, next) {
+    
+  }
+}
+
```
Comment (mrdivinemaniac) [inline server/middleware/index.js:6]: Our first custom middleware :D
We will implement this in the session.

File: server/security/index.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+function signToken (payload, expiresIn) {
+  return Promise.resolve('dummy.jwt.token')
+}
+
+function verifyToken (token) {
+  return Promise.resolve({
+    userId: 'someuserId'
+  })
+}
```
Comment (mrdivinemaniac) [inline server/security/index.js:14]: We will implement these in the session

File: server/security/index.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+function signAuthToken (user, scopes) {
+  return signToken({
+    userId: user.id,
+    scopes: Array.isArray(scopes) ? scopes : [scopes]
+  })
+}
```
Comment (mrdivinemaniac) [inline server/security/index.js:21]: scopes will be explained in the session

File: server/security/index.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+const jwt = require('jsonwebtoken')
+const { JWT_SECRET , JWT_ISSUER } = require('../config')
```
Comment (mrdivinemaniac) [inline server/security/index.js:2]: These are defined in a later commit. They will be explained in the session.

File: server/routes/posts/index.js [hunk 7]
```diff
@@ -43,6 +44,38 @@ router.post('/', upload.single('codeFile'), async (req, res, next) => {
...
+})
+
+router.patch('/:id', upload.single('codeFile'), async (req, res, next) => {
+  try {
+
+  } catch (e) {
+    next(e)
+  }
+})
+
```
Comment (mrdivinemaniac) [inline server/routes/posts/index.js:78]: We will implement these in the session.

File: server/routes/posts/index.js [hunk 8]
```diff
@@ -51,4 +84,7 @@ router.use((err, req, res, next) => {
   }
 })
 
+// TODOS Figure out how to retrieve postId in likesRouter
+router.use('/:postId/likes', likesRouter)
+
```
Comment (mrdivinemaniac) [inline server/routes/posts/index.js:89]: This will be one of the things you will try out yourself after the session

File: cli/utils/index.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+const path = require('path')
+const fetch = require('node-fetch')
+const config = require('../config')
+
+function requestApi (apiPath, ...rest) {
+  return fetch(`${config.apiHost}/${apiPath}`, ...rest)
+}
+
+function requestAuthenticatedApi (apiPath, options, ...rest) {
+  const token = readToken()
```
Comment (shrestharosy028) [inline cli/utils/index.js:25]: Where's readToken declared? Implement huna baki ho?

Comment (mrdivinemaniac) [reply cli/utils/index.js:25]: Yes! Jumbled up the commits. sorry!

File: cli/utils/index.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+      ...(options.headers || {})
+    }
+  }
+  return requestApi(apiPath, patchedOptions, ...rest)
+}
+
+module.exports = {
+  saveToken,
+  readToken,
+  deleteToken,
```
Comment (mrdivinemaniac) [inline cli/utils/index.js:40]: These are declared in the next commit

File: server/routes/auth/index.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
+const express = require('express')
+const revokedTokens = require('../../models/revoked-tokens')
+const users = require('../../models/users')
+const security = require('../../security')
+
+const router = express.Router()
+
+router.post('/token', express.json(), async (req, res, next) => {
```
Comment (mrdivinemaniac) [inline server/routes/auth/index.js:8]: `express.json` will be explained in the session.

**Pull request merged**